### PR TITLE
Package ppx_mica.0.1.0

### PR DESCRIPTION
**Pull request by ngernest**
### `ppx_mica.0.1.0`
PPX deriver that automates differential testing for OCaml modules

---
* Homepage: https://github.com/ngernest/mica
* Source repo: git+https://github.com/ngernest/mica.git
* Bug tracker: https://github.com/ngernest/mica/issues

---
:camel: Pull-request generated by opam-publish v2.4.0

### CODE REVIEW

**Comment by shonfeder** [comment]
Hello! Thanks for you for publishing your package Mica looks super cool, and the readme is extremely clear.

Since it looks like this is your first time publishing a package to opam, allow me offer a warm welcome and a bit of an intro :)

We run a CI system to ensure that packages can be installed on all intended systems. You can see the results for your PR at https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/51b56fad1033fa4ad04b706eef5074af6f223397 (or by clicking the `Details` link next to the `opam-ci` status check).

This shows the results of the builds and tests across our [build matrix](https://github.com/ocurrent/opam-repo-ci/blob/master/doc/platforms.md). Because we build and test on such a large matrix, it is very common for errors to be identified during package publication.

In my following comments and suggestions, I'll try to provide pointers and suggestions to get everything squared away. Please let us know if you have any questions, as we are here to help :)

File: packages/tyche_utils/tyche_utils.0.1.0/opam [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+synopsis:
+  "Utilities for Tyche to visualize results produced from tests run by Mica"
+maintainer: ["Ernest Ng <ernest@cs.cornell.edu>"]
+authors: ["Ernest Ng <ernest@cs.cornell.edu>"]
+license: "ISC"
+homepage: "https://github.com/ngernest/mica"
+bug-reports: "https://github.com/ngernest/mica/issues"
+depends: [
+  "dune" {>= "3.7"}
+  "yojson"
```
Comment (shonfeder) [inline packages/tyche_utils/tyche_utils.0.1.0/opam:12]: One of the errors is a "lower bounds check failure". The actual errors looks like this:

```
#=== ERROR while compiling tyche_utils.0.1.0 ==================================#
# context              2.3.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.11.2 | pinned(https://github.com/ngernest/mica/archive/refs/tags/v0.1.0.tar.gz)
# path                 ~/.opam/4.11/.opam-switch/build/tyche_utils.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p tyche_utils -j 39 @install
# exit-code            1
# env-file             ~/.opam/log/tyche_utils-7-a31302.env
# output-file          ~/.opam/log/tyche_utils-7-a31302.out
### output ###
# File "/home/opam/.opam/4.11/lib/ocaml-compiler-libs/bytecomp/ocaml-compiler-libs.bytecomp.dune", line 1, characters 0-0:
<snip>
# File "lib/tyche_utils/tyche_utils.ml", line 18, characters 13-27:
# 18 |   val json : Yojson.Basic.t
#                   ^^^^^^^^^^^^^^
# Error: Unbound type constructor Yojson.Basic.t
```

This is taken from the CI log https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/51b56fad1033fa4ad04b706eef5074af6f223397/variant/compilers,4.11,tyche_utils.0.1.0,lower-bounds

It indicates that the package depends on a feature of `Yojson` that is not available in some versions matching the dependency specification.

To fix this, we can add a lower bound. This is probably a safe bet, given the change log:

```suggestion
  "yojson" {>= "2.0.O"}
```

File: packages/tyche_utils/tyche_utils.0.1.0/opam [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  "base" {>= "v0.15.1"}
+  "core_unix" {>= "v0.15.0"}
+  "odoc" {with-doc}
```
Comment (shonfeder) [inline packages/tyche_utils/tyche_utils.0.1.0/opam:15]: We have [tests (failed: Library "alcotest" not found.)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/51b56fad1033fa4ad04b706eef5074af6f223397/variant/compilers,4.11,tyche_utils.0.1.0,tests)

```
#=== ERROR while compiling tyche_utils.0.1.0 ==================================#
# context              2.3.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.11.2 | pinned(https://github.com/ngernest/mica/archive/refs/tags/v0.1.0.tar.gz)
# path                 ~/.opam/4.11/.opam-switch/build/tyche_utils.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p tyche_utils -j 39 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/tyche_utils-7-ad7cb0.env
# output-file          ~/.opam/log/tyche_utils-7-ad7cb0.out
### output ###
# File "test/utils_test/dune", line 3, characters 21-29:
# 3 |  (libraries ppx_mica alcotest ppxlib)
#                          ^^^^^^^^
# Error: Library "alcotest" not found.
# -> required by _build/default/test/utils_test/test_runner.exe
# -> required by alias test/utils_test/runtest in test/utils_test/dune:2
```

Which is just indicating that the package doesn't declare the needed test dependency on alcotest:

```suggestion
  "odoc" {with-doc}  
  "alcotest" {with-test}
```

File: packages/mica_case_studies/mica_case_studies.0.1.0/opam [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# This file is generated by dune, edit dune-project instead
+opam-version: "2.0"
+synopsis: "Mica case studies (for OCaml Workshop 2024 talk)"
```
Comment (shonfeder) [inline packages/mica_case_studies/mica_case_studies.0.1.0/opam:3]: Is this package expected to be of value for end-users? If it is only/primarily meant as an example or for test cases, please remove it from the publication, as per our policy that [Packages should provide some sort of utility for users](https://github.com/ocaml/opam-repository/wiki/Policies#3-packages-should-provide-some-sort-of-utility-for-users). However, if it is of some general utility, please add a description to explain.

File: packages/ppx_mica/ppx_mica.0.1.0/opam [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+bug-reports: "https://github.com/ngernest/mica/issues"
+depends: [
+  "dune" {>= "3.7"}
+  "ocaml" {>= "4.13"}
+  "odoc" {with-doc}
+  "core" {>= "v0.15.1"}
+  "ppx_jane" {>= "v0.15.0"}
+  "ppx_assert" {>= "v0.15.0"}
+  "base" {>= "v0.15.1"}
+  "base_quickcheck" {>= "v0.15.0"}
```
Comment (shonfeder) [inline packages/ppx_mica/ppx_mica.0.1.0/opam:17]: [ppx_mica.0.1.0 (failed: Library "tyche_utils" not found.)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/51b56fad1033fa4ad04b706eef5074af6f223397/variant/compilers,4.13,ppx_mica.0.1.0) tells us that we are missing the dependency on the `tyche_utils` here:

```suggestion
  "base_quickcheck" {>= "v0.15.0"}
  "tyche_utils" {=version}
```

This uses the [version package variable](https://opam.ocaml.org/doc/Manual.html#pkgvar-version) to ensure the versions stay in sync. If you expect the API of the utils to be stable accross versions, you could probably omit this constraint.

**Review by shonfeder — COMMENTED** [review]
I think this should address all the errors I am seeing!

File: packages/mica_case_studies/mica_case_studies.0.1.0/opam (hunk 3 above)
Comment (ngernest) [reply packages/mica_case_studies/mica_case_studies.0.1.0/opam]: Ah I see, thanks for the heads-up! I've removed this package (it's now been isolated in a [separate repo](https://github.com/ngernest/mica_case_studies)) and its associated `.opam` file.

File: packages/ppx_mica/ppx_mica.0.1.0/opam (hunk 4 above)
Comment (ngernest) [reply packages/ppx_mica/ppx_mica.0.1.0/opam:17]: Thanks! For simplicity, I've moved the `tyche_utils` package to a separate repo as it's not strictly necessary for the main `ppx_mica` package.

File: packages/tyche_utils/tyche_utils.0.1.0/opam (hunk 1 above)
Comment (ngernest) [reply packages/tyche_utils/tyche_utils.0.1.0/opam]: Thanks for the info, very helpful!

File: packages/tyche_utils/tyche_utils.0.1.0/opam (hunk 2 above)
Comment (ngernest) [reply packages/tyche_utils/tyche_utils.0.1.0/opam]: Thanks!

**Comment by ngernest** [comment]
Thanks @shonfeder for the comments! I found them very helpful as a first-time package author. 

I've edited the PR so that there's only one package (`ppx_mica`) and added lower bounds for its dependencies -- the other two packages (`tyche_utils` & `mica_case_studies`) aren't strictly necessary for end users, so I've removed them from this PR and isolated them in separate repos.

File: packages/ppx_mica/ppx_mica.0.1.0/opam [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  "ppx_deriving" {>= "6.0.2"}
+  "ppxlib" {>= "0.33.0"}
+  "alcotest" {>= "1.8.0"}
```
Comment (shonfeder) [inline packages/ppx_mica/ppx_mica.0.1.0/opam:20]:
```suggestion
  "alcotest" {with-test & >>= "1.8.0"}
```

File: packages/ppx_mica/ppx_mica.0.1.0/opam [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+]
```
Comment (shonfeder) [inline packages/ppx_mica/ppx_mica.0.1.0/opam:21]:
```suggestion
  "core_unix" {>= "v0.17.0"}
]
```

**Review by shonfeder — COMMENTED** [review]
Thank you for the fixes and the package clean up! Things are looking quite good :)

There is one more [error turning up in the ci](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/07a5b35b33983c2baf74de22bdea1d1134726c30) dune to a missing dependency.

I opened https://github.com/ngernest/mica/pull/3 to fix this in the source, and the suggestions here should take care of this package (tho pushing an update from the opam file your dev repo is a fine alternative to adding the changes here directly).

Looking forward to having this in the repo! Today I was day dreaming about whether we could use your work here in our CI system somehow to detect breaking changes between releases :D

**Comment by ngernest** [comment]
Thanks for the fixes & for the PR (I merged your PR)! I pushed an update from the Opam file in my dev repo (I added a documentation stanza in my `dune-project` file) and incorporated your changes, hopefully this works!

**Comment by shonfeder** [comment]
It's working! Everything is coming up green. Thanks very much for your patience in the process and for publishing the package!

Done forget to announce in discuss.ocaml.org :)

**Comment by ngernest** [comment]
@shonfeder Thank you so much for all your help! I learned a lot and really enjoyed the process!

**Pull request merged**